### PR TITLE
Scene Graph Image Tutorial

### PR DESCRIPTION
**Pull request by paroma**
Add scene graph image tutorial, where task is to classify whether the relationship `riding` exists between a pair of bounding boxes. 

- [x] Port over old tutorial
- [x] Change dataset to VRD from VG
- [x] Data Preprocessing
     - [x] Filter Semantic Relationships
     - [x] Convert from per-image to per-relationship labels
     - [x] Create Pandas data frame for train, test, and val sets
- [x] Test LFs
- [x] Test LabelModel
- [x] Incorporate EndModel (done by @vincentschen)
     - [x] Choose architecture (TripleCNN?)
     - [x] Train Model (save pertained version?)
- [x] Write scripts 
     - [x] Preprocessing script
     - [x] Download script
     - [x] End Model script
- [x] Add instructions for understanding workflow
- [x] Change to multiclass

### CODE REVIEW

File: scene_graph/vrd_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    if x.subject_category == "person":
+        if x.object_category in ["bike", "snowboard", "motorcycle", "horse"]:
+            return RIDE
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_ride_rare_object(x):
+    if x.subject_category == "person":
+        if x.object_category in ["bus", "truck", "elephant"]:
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: 🐘

File: scene_graph/vrd_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+# %%
+from scene_graph.utils import load_vrd_data
+
+train_df, valid_df, test_df = load_vrd_data()
+
+print("Train Relationships: ", len(train_df))
+print("Dev Relationships: ", len(valid_df))
+print("Test Relationships: ", len(test_df))
+
+# %% [markdown]
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: might be tricky given our image download woes—but could be cool to look at a few examples up before this!

File: scene_graph/vrd_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+
+
+@labeling_function()
+def LF_person(x):
+    if x.subject_category != "person":
+        return OTHER
+    return ABSTAIN
+
+
+# Distance-based LFs
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: let's separate each set of LFs and write a sentence describing them?

File: scene_graph/vrd_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+
+
+@labeling_function()
+def LF_dist(x):
+    if np.linalg.norm(np.array(x.subject_bbox) - np.array(x.object_bbox)) <= 1000:
+        return OTHER
+    return ABSTAIN
+
+
+# Size-based LF
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: separate!

File: scene_graph/vrd_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+        x.object_bbox[3] - x.object_bbox[2]
+    )
+
+    if subject_area / object_area <= 0.5:
+        return OTHER
+    return ABSTAIN
+
+
+# %% [markdown]
+# Note that the labeling functions have varying empirical accuracies and coverages. Because of the imbalance in how the classes are defined, labeling functions that label the `OTHER` class have higher coverage than labeling functions for `RIDE` or `CARRY`. This reflects the distribution of classes in the dataset as well.
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: > Because of the imbalance in how the classes are defined

_Due to class imbalance in our chosen relationships_

File: scene_graph/vrd_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+from snorkel.labeling.apply import PandasLFApplier
+lfs = [
+    LF_ride_object,
+    LF_ride_rare_object,
+    LF_carry_object,
+    LF_carry_subject,
+    LF_person,
+    LF_ydist,
+    LF_dist,
+    LF_area,
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: nit: extra comma

File: scene_graph/vrd_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+L_valid = applier.apply(valid_df)
+
+# %%
+from snorkel.labeling.analysis import LFAnalysis
+
+Y_valid = valid_df.label.values
+lf_names = [lf.name for lf in lfs]
+
+lf_analysis = LFAnalysis(L_valid)
+lf_analysis.lf_summary(Y_valid, lf_names=lf_names)
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: nit: is it cleaner to do `LFAnalysis(L_valid).lf_summary(Y_valid, lf_names=lf_names)` ?

File: scene_graph/vrd_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+# We now train a multi-class `LabelModel` to assign training labels to the unalabeled training set.
+
+# %%
+from snorkel.labeling.model import LabelModel
+
+label_model = LabelModel(cardinality=3, verbose=True)
+label_model.fit(L_train, seed=123, lr=0.01, log_freq=10, n_epochs=100)
+
+# %%
+label_model.score(L_valid, Y_valid, metrics=["f1_micro"])
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: do you think it's important to briefly explain `f1_micro`? (given that we forgot which one that was, haha)

File: scene_graph/vrd_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+
+# %% [markdown]
+# ## 4. Train a Classifier
+# You can then use these training labels to train any standard discriminative model, such as [a state-of-the-art ResNet](https://github.com/KaimingHe/deep-residual-networks), which should learn to generalize beyond the LF's we've developed!
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: nit: I'm not sure if we can claim state-of-the-art for ResNet anymore

File: scene_graph/vrd_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+
+# %% [markdown]
+# #### Create DataLoaders for Classifier
+
+# %%
+from snorkel.classification.data import DictDataLoader
+from scene_graph.model import FlatConcat, SceneGraphDataset, WordEmb, init_fc
+
+#change to "scene_graph/data/VRD/sg_dataset/sg_train_images" for full set
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: suggestion to clean up: we can set this based on environment variable `os.environ["TRAVIS"]` in utils.py, and then `import TRAIN_DIR`

File: scene_graph/vrd_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+)
+
+# %% [markdown]
+# #### Define Loss Functions
+
+
+# %%
+import torch.nn.functional as F
+
+def ce_loss(module_name, outputs, Y, active):
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: we can actually import these from `from snorkel.classification.task import ce_loss, softmax`

File: scene_graph/vrd_tutorial.py [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+    return F.cross_entropy(outputs[module_name][0][active], (Y.view(-1))[active])
+
+
+def softmax(module_name, outputs):
+    return F.softmax(outputs[module_name][0], dim=1)
+
+
+# %% [markdown]
+# #### Define Model Architecture
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: reminder to make a pointer to the MTL tutorial here "for more" once that is merged

File: scene_graph/vrd_tutorial.py [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+# define layers
+module_pool = nn.ModuleDict(
+    {
+        "feat_extractor": feature_extractor,
+        "prediction_head": fc,
+        "feat_concat": FlatConcat(),
+        "word_emb": WordEmb(),
+    }
+)
+
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: comment: # define feature extractors for each of the (union, subject, and object) image crops

File: scene_graph/vrd_tutorial.py [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+sub_feat_op = Operation(
+    name="sub_feat_op", module_name="feat_extractor", inputs=[("_input_", "sub_crop")]
+)
+
+obj_feat_op = Operation(
+    name="obj_feat_op", module_name="feat_extractor", inputs=[("_input_", "obj_crop")]
+)
+
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: comment: # define an operation to extract word embeddings for subject and object categories

File: scene_graph/vrd_tutorial.py [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+word_emb_op = Operation(
+    name="word_emb_op",
+    module_name="word_emb",
+    inputs=[("_input_", "sub_category"), ("_input_", "obj_category")],
+)
+
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: comment: # define an operation to concatenate image features and word embeddings

File: scene_graph/vrd_tutorial.py [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+    name="concat_op",
+    module_name="feat_concat",
+    inputs=[
+        ("obj_feat_op", 0),
+        ("sub_feat_op", 0),
+        ("union_feat_op", 0),
+        ("word_emb_op", 0),
+    ],
+)
+
```
Comment (vincentschen) [inline scene_graph/vrd_tutorial.py]: comment: # define an operation to make a prediction over all concatenated features

File: scene_graph/model.py [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+            "sub_bbox": df["subject_bbox"].tolist(),
+            "obj_category": df["object_category"].tolist(),
+            "sub_category": df["subject_category"].tolist(),
+        }
+        Y_dict = {
+            "scene_graph_task": torch.LongTensor(df["label"].to_numpy())
+        }  # change to take in the rounded train labels
+        super(SceneGraphDataset, self).__init__(name, split, X_dict, Y_dict)
+
+        # define standard set of transformations to apply to each image
```
[the remark this answers is not in the document]
Comment (paroma) [reply scene_graph/model.py]: we could actually have TFs that add labels. for example, every "person riding bike" is also a "bike carrying person"

File: scene_graph/utils.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+
+
+# %%
+def vrd_to_pandas(
+    relationships_set, objects, predicates, list_of_predicates, keys_list=None
+):
+    relationships = []
+
+    for img in relationships_set:
+        if (keys_list is None) or (img in keys_list):
```
[the remark this answers is not in the document]
Comment (paroma) [reply scene_graph/utils.py]: if we're using the full dataset, I don't pass the keys around (since there's two separate directories in that case)

File: scene_graph/vrd_tutorial.py (hunk 3 above)
Comment (paroma) [reply scene_graph/vrd_tutorial.py]: I added paragraphs above each set instead of each LF, they're not too complex

File: scene_graph/vrd_tutorial.py (hunk 7 above)
Comment (paroma) [reply scene_graph/vrd_tutorial.py]: probably but in case users want to run and add their own fields they want to calculate, that usage isn't recommended

File: scene_graph/vrd_tutorial.py (hunk 9 above)
Comment (paroma) [reply scene_graph/vrd_tutorial.py]: replaced with off-the-shelf

File: scene_graph/model.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (vincentschen) [reply scene_graph/model.py]: nice! non-goal for this push, but noted

File: scene_graph/vrd_tutorial.py (hunk 7 above)
Comment (vincentschen) [reply scene_graph/vrd_tutorial.py]: you're right, sounds god
